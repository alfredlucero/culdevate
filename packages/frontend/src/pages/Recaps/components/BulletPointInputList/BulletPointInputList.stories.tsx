import React, { useState } from "react";
import { DropResult } from "react-beautiful-dnd";
import { storiesOf } from "@storybook/react";
import BulletPointInputList, { BulletPointInputListItem } from "./index";

const StatefulBulletPointInputList = () => {
  const [bulletPointInputList, setBulletPointInputList] = useState<BulletPointInputListItem[]>([
    {
      value: "",
      id: "bulletpoint-0",
      errorInfo: "",
      valid: true,
    },
  ]);

  const handleChangeBulletPointInput = (e: React.FormEvent<HTMLInputElement>) => {
    const { value, id } = e.currentTarget;

    const newBulletPointList = bulletPointInputList.map(bulletPointInputListItem => {
      if (bulletPointInputListItem.id === id) {
        return {
          ...bulletPointInputListItem,
          value,
        };
      }

      return bulletPointInputListItem;
    });

    setBulletPointInputList(newBulletPointList);
  };

  const handleBlurBulletPointInput = (e: React.FocusEvent<HTMLInputElement>) => {
    const { id, value } = e.currentTarget;

    const newBulletPointList = bulletPointInputList.map(bulletPointInputListItem => {
      if (bulletPointInputListItem.id === id) {
        const errorInfo = value === "" ? "Bullet point is required." : "";
        const valid = errorInfo === "";
        return {
          ...bulletPointInputListItem,
          valid,
          errorInfo,
        };
      }
      return bulletPointInputListItem;
    });

    setBulletPointInputList(newBulletPointList);
  };

  const handleDeleteBulletPointInput = (e: React.MouseEvent) => {
    const { id } = e.currentTarget;

    const newBulletPointInputList = bulletPointInputList
      .filter(bulletPointInputListItem => {
        return bulletPointInputListItem.id !== id;
      })
      .map((bulletPointInputListItem, index) => ({
        ...bulletPointInputListItem,
        id: `bulletpoint-${index}`,
      }));

    setBulletPointInputList(newBulletPointInputList);
  };

  const handleAddBulletPointInput = () => {
    const newBulletPointInputList = [
      ...bulletPointInputList,
      {
        value: "",
        id: `bulletpoint-${bulletPointInputList.length}`,
        errorInfo: "",
        valid: true,
      },
    ];

    setBulletPointInputList(newBulletPointInputList);
  };

  const reorderBulletPointInputList = (
    bulletPointInputList: BulletPointInputListItem[],
    startIndex: number,
    endIndex: number,
  ): BulletPointInputListItem[] => {
    const currentBulletPointInputList = Array.from(bulletPointInputList);
    const [removedBulletPointIndex] = currentBulletPointInputList.splice(startIndex, 1);
    currentBulletPointInputList.splice(endIndex, 0, removedBulletPointIndex);

    return currentBulletPointInputList;
  };

  const handleDragEnd = (dropResult: DropResult) => {
    const isDroppedOutsideList = !dropResult.destination;
    if (isDroppedOutsideList) {
      return;
    }

    if (dropResult.source && dropResult.destination) {
      const isDroppedAtSamePlaceInList = dropResult.destination.index === dropResult.source.index;
      if (isDroppedAtSamePlaceInList) {
        return;
      }

      const reorderedBulletPointInputList = reorderBulletPointInputList(
        bulletPointInputList,
        dropResult.source.index,
        dropResult.destination.index,
      );

      setBulletPointInputList(reorderedBulletPointInputList);
    }
  };

  return (
    <div>
      <p>Bullet Point Input List</p>
      <BulletPointInputList
        bulletPointInputList={bulletPointInputList}
        onDragEnd={handleDragEnd}
        onChangeBulletPointInput={handleChangeBulletPointInput}
        onBlurBulletPointInput={handleBlurBulletPointInput}
        onDeleteBulletPointInput={handleDeleteBulletPointInput}
        onAddBulletPointInput={handleAddBulletPointInput}
        isAddBulletPointInputDisabled={false}
      />
    </div>
  );
};

storiesOf("RecapsPage/BulletPointInputList", module).add("Stateful Example", () => <StatefulBulletPointInputList />);
