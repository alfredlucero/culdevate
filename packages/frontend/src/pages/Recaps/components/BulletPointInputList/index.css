.bullet-point-list-droppable {
  min-height: 200px;
  max-height: 300px;
  overflow-y: auto;
}

.add-bullet-point-plus-icon {
  color: inherit;
}
