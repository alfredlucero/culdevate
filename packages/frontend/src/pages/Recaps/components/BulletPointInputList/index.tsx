import React from "react";
import cn from "classnames";
import { DragDropContext, Droppable, Draggable, DropResult } from "react-beautiful-dnd";
import BulletPointInput, { BulletPointInputProps } from "../BulletPointInput";
import Button from "../../../../components/Button";
import Icon from "../../../../components/Icon";
import { CommonProps } from "../../../../components/commonProps";
import "./index.css";

export interface BulletPointInputListProps extends CommonProps {
  bulletPointInputList: BulletPointInputListItem[];
  onDragEnd: (dropResult: DropResult) => void;
  onChangeBulletPointInput: BulletPointInputProps["onChange"];
  onBlurBulletPointInput: BulletPointInputProps["onBlur"];
  onDeleteBulletPointInput: BulletPointInputProps["onDelete"];
  onAddBulletPointInput: () => void;
  isAddBulletPointInputDisabled: boolean;
}

export interface BulletPointInputListItem {
  value: string;
  id: string;
  errorInfo: string;
  valid: boolean;
}

const BulletPointInputList: React.FC<BulletPointInputListProps> = ({
  bulletPointInputList,
  onDragEnd,
  onChangeBulletPointInput,
  onBlurBulletPointInput,
  onDeleteBulletPointInput,
  onAddBulletPointInput,
  isAddBulletPointInputDisabled,
  className = "",
  testId = "",
  ...passThroughProps
}) => {
  return (
    <DragDropContext onDragEnd={onDragEnd}>
      <div className={cn(className)} {...(testId !== "" ? { "data-testid": testId } : {})} {...passThroughProps}>
        <Droppable droppableId="bulletPointInputList">
          {(provided, snapshot) => (
            <div
              {...provided.droppableProps}
              ref={provided.innerRef}
              className={cn(
                {
                  "bg-gray-300": !snapshot.isDraggingOver,
                  "bg-green-200": snapshot.isDraggingOver,
                },
                "pt-6",
                "px-4",
                "pb-4",
                "bullet-point-list-droppable",
              )}
            >
              {bulletPointInputList.map((bulletPointInputListItem, index) => (
                <Draggable key={bulletPointInputListItem.id} draggableId={bulletPointInputListItem.id} index={index}>
                  {(provided, snapshot) => (
                    <div
                      ref={provided.innerRef}
                      {...provided.draggableProps}
                      {...provided.dragHandleProps}
                      className={cn(
                        {
                          "bg-white": !snapshot.isDragging,
                          "bg-blue-200": snapshot.isDragging,
                        },
                        "hover:bg-blue-100",
                        "rounded",
                      )}
                    >
                      <BulletPointInput
                        onChange={onChangeBulletPointInput}
                        onBlur={onBlurBulletPointInput}
                        onDelete={onDeleteBulletPointInput}
                        value={bulletPointInputListItem.value}
                        id={bulletPointInputListItem.id}
                        valid={bulletPointInputListItem.valid}
                        errorInfo={bulletPointInputListItem.errorInfo}
                        className="mb-2"
                      />
                    </div>
                  )}
                </Draggable>
              ))}
              {provided.placeholder}
              <div className="flex justify-center">
                <Button
                  variant="secondary"
                  type="button"
                  disabled={isAddBulletPointInputDisabled}
                  onClick={onAddBulletPointInput}
                  className="mt-4 flex items-center"
                >
                  <Icon variant="plus" size="small" className="mr-2 add-bullet-point-plus-icon" />
                  Add Bullet Point
                </Button>
              </div>
            </div>
          )}
        </Droppable>
      </div>
    </DragDropContext>
  );
};

export default BulletPointInputList;
