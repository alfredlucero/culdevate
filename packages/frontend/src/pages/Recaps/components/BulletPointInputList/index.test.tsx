import React from "react";
import { render } from "@testing-library/react";
import BulletPointInputList, { BulletPointInputListProps } from "./index";

describe("<BulletPointInputList />", () => {
  test("should render without error", () => {
    const bulletPointInputList = [
      {
        value: "",
        id: "bulletpoint-0",
        errorInfo: "",
        valid: true,
      },
    ];

    const { container } = render(
      <BulletPointInputList
        bulletPointInputList={bulletPointInputList}
        onDragEnd={() => {}}
        onChangeBulletPointInput={() => {}}
        onBlurBulletPointInput={() => {}}
        onDeleteBulletPointInput={() => {}}
        onAddBulletPointInput={() => {}}
        isAddBulletPointInputDisabled={false}
        className="extra-input-list-class"
        testId="bulletPointInputList"
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
