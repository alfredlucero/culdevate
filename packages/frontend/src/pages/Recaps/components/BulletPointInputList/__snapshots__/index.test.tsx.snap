// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BulletPointInputList /> should render without error 1`] = `
<div>
  <div
    class="extra-input-list-class"
    data-testid="bulletPointInputList"
  >
    <div
      class="bg-gray-300 pt-6 px-4 pb-4 bullet-point-list-droppable"
      data-rbd-droppable-context-id="0"
      data-rbd-droppable-id="bulletPointInputList"
    >
      <div
        aria-labelledby="rbd-lift-instruction-0"
        class="bg-white hover:bg-blue-100 rounded"
        data-rbd-drag-handle-context-id="0"
        data-rbd-drag-handle-draggable-id="bulletpoint-0"
        data-rbd-draggable-context-id="0"
        data-rbd-draggable-id="bulletpoint-0"
        draggable="false"
        tabindex="0"
      >
        <div
          class="flex items-start px-4 py-6 rounded shadow-md cursor-pointer mb-2"
        >
          <div
            class="font-sans text-gray-800 text-2xl mr-4 text-teal-500"
            id=""
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-circle-right fa-w-16 "
              data-icon="chevron-circle-right"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zm113.9 231L234.4 103.5c-9.4-9.4-24.6-9.4-33.9 0l-17 17c-9.4 9.4-9.4 24.6 0 33.9L285.1 256 183.5 357.6c-9.4 9.4-9.4 24.6 0 33.9l17 17c9.4 9.4 24.6 9.4 33.9 0L369.9 273c9.4-9.4 9.4-24.6 0-34z"
                fill="currentColor"
              />
            </svg>
          </div>
          <div
            class="mr-4 flex-1"
          >
            
            <input
              aria-required="true"
              class="shadow appearance-none border w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline rounded"
              id="bulletpoint-0"
              name="bulletpoint-0"
              placeholder="Provide your highlights and results..."
              type="text"
              value=""
            />
          </div>
          <div
            class="font-sans text-gray-800 text-2xl text-red-500 hover:text-red-700 cursor-pointer"
            id="bulletpoint-0"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-trash-alt fa-w-14 "
              data-icon="trash-alt"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="flex justify-center"
    >
      <button
        class="font-bold py-2 px-4 rounded bg-transparent text-teal-700 border border-teal-500 hover:bg-teal-500 hover:text-white hover:border-transparent mt-4 flex items-center"
        type="button"
      >
        <div
          class="font-sans text-gray-800 text-xl mr-2 add-bullet-point-plus-icon"
          id=""
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-plus fa-w-14 "
            data-icon="plus"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
              fill="currentColor"
            />
          </svg>
        </div>
        Add Bullet Point
      </button>
    </div>
  </div>
</div>
`;
